src/api/libcellml/error.h

```diff
@@ -122,6 +122,20 @@ class LIBCELLML_EXPORT Error
         XML
     };
 
+    /**
+     * @brief The error Level enum class
+     * 
+     * Enum to describe the level of the error
+     */
+    enum class Level
+    {
+        UNDEFINED,
+        FATAL,
+        WARNING,
+        FUTURE_ERROR,
+        INFORMATION
```

Comment:
1. I guess the class is going to be renamed from `Error` to `Issue` at some point, and that the `.cpp` and `.h` files are going to be renamed too?
2. Can we and do we really want an `UNDEFINED` issue? Also, what is a `FUTURE_ERROR`? I would just have the type (not level) of issue that we think we should have in the end, which I imagine are: `ERROR`, `WARNING` and `HINT`?

Reply:
1. Agreed ... but I've shelved this until all the other outstanding PRs are merged in as there's just no point in doing it all twice!
2. Agreed as well - the default is `FATAL` (at present) to be backward consistent.  It was there because there was an `UNDEFINED` in the other structures and I was following along ... 
FUTURE_ERROR was something I was wondering about - we could use it two ways.  A lot of packages have deprecation warnings, so it could either be that.  But what I was actually heading for was "you have something which is allowed in the parser but will not be valid" or "you have something which is valid but which will not actually run or mean anything", ie: something which conforms to the specs and rules but will create a problem further down the pipeline.  Not much difference from a warning, really ...

Reply:
Others may chime in, but:

1. How do you mean no point in doing it all twice? Are you referring to the fact that other PRs currently rely on the `Error` class? Either way, this issue is about reworking the `Error` class. So, I would expect this PR to be a full rework, not a half-baked one.
2. Why should we be backward compatible? There is no official release of libCellML, so it's not like we are going to break something. Otherwise, I am not convinced by `FATAL` and `UNDEFINED`. `FATAL` is subjective, so will leave the others to comment, but `UNDEFINED` is just wrong to me. libCellML should know what kind of issue it reports. As for `FUTURE_ERROR`, your first use could be addressed with a `WARNING` issue. As for your second possible use, I don't think the parser should worry about what is allowed, but is not valid. This will be the job of the model analyser to determine whether a valid model (in the parser sense) makes sense or not.

Reply:
> "Undefined is wrong to me" 

What's the rationale of having it elsewhere?  I was just following what's already there.  Do we need to remove `UNDEFINED` in general then?

> Fatal is subjective

I don't understand ... Currently all our errors are effectively fatal errors hence use of that as a default?

> This will be the job of the model analyser to determine whether a valid model (in the parser sense) makes sense or not.

Maybe I'm not understanding the purpose of this class the same way everyone else is - I'd imagined that it was to better support model analysis and debugging by returning to the user (when asked) where possible glitches and errors might occur, based on information that libcellml currently knows (like stuff from the parser/validator/imports) but doesn't currently report?

> How do you mean no point in doing it all twice? Are you referring to the fact that other PRs currently rely on the Error class? Either way, this issue is about reworking the Error class. So, I would expect this PR to be a full rework, not a half-baked one.

Fully-baked is intended :) ... but it's also (nudge nudge) been my experience that PRs take rather a long time to be merged in, which means that their fix-merge-conflicts-and-design-reworks-factor-to-accommodate-work-done-since-writing becomes pretty high.  There's no point - for example - in going through the current errors in development regarding resets, validation, parsing, whens, etc as these won't exist as soon as the outstanding PRs are merged in - it's wasted work.  (If that sounds overly narky I apologise - today, the sarcasm is strong with this one!)

Reply:
> > "Undefined is wrong to me"
> What's the rationale of having it elsewhere? I was just following what's already there. Do we need to remove `UNDEFINED` in general then?

When it comes to a libCellML issue, I would remove the `UNDEFINED` type since an undefined issue doesn't make sense to me. There may be other cases where an `UNDEFINED` type might be necessary. For example, we might need such a type when trying to determine the type of a variable in a model (i.e. state, constant, computed constant, etc. or undefined indeed, if we couldn't determine the type of the variable, e.g. the model is under/overconstrained).

> > Fatal is subjective
> I don't understand ... Currently all our errors are effectively fatal errors hence use of that as a default?

Sorry, I meant that the naming of the type itself is subjective. IOW, I would personally have `ERROR` rather than `FATAL`.

> > This will be the job of the model analyser to determine whether a valid model (in the parser sense) makes sense or not.
> 
> Maybe I'm not understanding the purpose of this class the same way everyone else is - I'd imagined that it was to better support model analysis and debugging by returning to the user (when asked) where possible glitches and errors might occur, based on information that libcellml currently knows (like stuff from the parser/validator/imports) but doesn't currently report?

For me (others may have a different view), the `Issue` class is to be used to report any kind of issue that we come across, be it either as a result of parsing a model, validating a model, analysing a model, etc. In that context, I feel like we need to be able to specify whether an issue is an error, a warning or a hint (any other relevant type?).

> > How do you mean no point in doing it all twice? Are you referring to the fact that other PRs currently rely on the Error class? Either way, this issue is about reworking the Error class. So, I would expect this PR to be a full rework, not a half-baked one.
> 
> Fully-baked is intended :) ... but it's also (nudge nudge) been my experience that PRs take rather a long time to be merged in, which means that their fix-merge-conflicts-and-design-reworks-factor-to-accommodate-work-done-since-writing becomes pretty high. There's no point - for example - in going through the current errors in development regarding resets, validation, parsing, whens, etc as these won't exist as soon as the outstanding PRs are merged in - it's wasted work. (If that sounds overly narky I apologise - today, the sarcasm is strong with this one!)

Trust me, I do sympathise with your sentiment with regards to the merging of PRs... :)

Reply:
> There may be other cases where an UNDEFINED type might be necessary.

I'm not talking about in the maths part of the code - it's in the existing `Error::Kind` enum (see below).  Personally I think that having a meaningful default (probably fatal/error/you-shall-not-pass/whatever) is better than undefined, but was trying to be consistent ;) . BTW the fact that there's a "kind" means I don't really want to have a "type" as well - they're too close in implication!  I think "level" or "severity" or something which implies effect or outcome (cf "kind" which implies cause) is clearer? 

```
    /**
     * @brief The error Kind enum class.
     *
     * Enum to describe the kind of error a given error is.
     */
    enum class Kind
    {
        COMPONENT,
            ... 
        RESET,
        UNDEFINED,  // <<<<<<<<<
        UNITS,
        VARIABLE,
        WHEN,
        XML
    };

// also in specificationrules.h:

/**
 * @brief The SpecificationRule enum class.
 *
 * Enum describing the specification rules that are used
 * for validation.
 */
enum class SpecificationRule
{
    UNDEFINED,  // <<<<<<<<<<<<<<<<<
    DATA_REPR_IDENTIFIER_UNICODE,
    DATA_REPR_IDENTIFIER_LATIN_ALPHANUM,
    DATA_REPR_IDENTIFIER_AT_LEAST_ONE_ALPHANUM,
        ...
```

> Sorry, I meant that the naming of the type itself is subjective. IOW, I would personally have ERROR rather than FATAL.

That's easy enough :)